storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -456,7 +457,21 @@ public void releaseSlotFor(LocalAssignment assignment, int port) throws IOExcept
             topoConfBlob.removeReference(pna);
         }
 
-        for (LocalResource lr : getLocalResources(pna)) {
+        List<LocalResource> localResources = null;
+        try {
+            // Precondition1: Base blob stormconf.ser and stormcode.ser have been localized
+            // Precondition2: Both these two blob files are fully downloaded and proper permission been set
+            localResources = getLocalResources(pna);
+        } catch (FileNotFoundException fnfException) {
+            LOG.warn("Local base blobs have not been downloaded yet. "
```

Comment:
can we add the pna to this message and the IOException error?  What is the side effect of not releasing this if there's a race condition?  Could the localizer then be updating this file continually?

Reply:
Added pna info. As for the race condition, I think that could be addressed by separating executor pools. I did not think of a scenario that race condition will happen because of catching the exception. 

All downloading tasks are CompletableFuture object. File no found happens when the object is not completed. So here we will forget about it, and dereference any resources we can, and release the slot for new assignments.

Then the background download threads will succeeds at some points. The cleanup thread will take care of these orphan files even if they are not released properly.